### PR TITLE
Add timestamp to events

### PR DESCRIPTION
**Pull request by windmemory**
Refer to https://github.com/Chatie/wechaty/issues/1829

### CODE REVIEW

File: src/user/room.ts [hunk 1]
```diff
@@ -543,10 +543,10 @@ export class Room extends Accessory implements Sayable {
     }
   }
 
-  public emit (event: 'invite', inviter: Contact,         invitation: RoomInvitation)           : boolean
-  public emit (event: 'leave',  leaverList:   Contact[],  remover?: Contact)                    : boolean
-  public emit (event: 'join',   inviteeList:  Contact[],  inviter:  Contact)                    : boolean
-  public emit (event: 'topic',  topic:        string,     oldTopic: string,   changer: Contact) : boolean
+  public emit (event: 'invite', inviter: Contact,         invitation: RoomInvitation)                       : boolean
+  public emit (event: 'leave',  time: Date, leaverList:   Contact[],  remover?: Contact)                    : boolean
```
Comment (huan) [inline src/user/room.ts]: Please move all the `time` argument to the end of the argument lists, and make it optional, so that it will not become a breaking change.

Comment (huan) [reply src/user/room.ts]: And also, please change the `remover?` argument for the `emit` function to `remover`. I think it's better to not optional as a emit argument.

File: src/wechaty.ts [hunk 2]
```diff
@@ -311,19 +311,19 @@ export class Wechaty extends Accessory implements Sayable {
...
-  public on (event: 'room-invite',  listener: string | ((this: Wechaty, roomInvitation: RoomInvitation) => void))                                   : this
-  public on (event: 'room-join',    listener: string | ((this: Wechaty, room: Room, inviteeList: Contact[],  inviter: Contact) => void))            : this
-  public on (event: 'room-leave',   listener: string | ((this: Wechaty, room: Room, leaverList: Contact[], remover?: Contact) => void))             : this
-  public on (event: 'room-topic',   listener: string | ((this: Wechaty, room: Room, newTopic: string, oldTopic: string, changer: Contact) => void)) : this
-  public on (event: 'scan',         listener: string | ((this: Wechaty, qrcode: string, status: ScanStatus, data?: string) => void))                : this
-  public on (event: 'start' | 'stop', listener: string | ((this: Wechaty) => void))                                                                 : this
+  public on (event: 'dong',         listener: string | ((this: Wechaty, data?: string) => void))                                                                : this
```
Comment (huan) [inline src/wechaty.ts]: Please do not include unrelated changes.

File: src/wechaty.ts [hunk 3]
```diff
@@ -683,13 +683,13 @@ export class Wechaty extends Accessory implements Sayable {
             const inviter = this.Contact.load(inviterId)
             await inviter.ready()
 
-            this.emit('room-join', room, inviteeList, inviter)
-            room.emit('join', inviteeList, inviter)
+            this.emit('room-join', room, new Date(timestamp), inviteeList, inviter)
```
Comment (huan) [inline src/wechaty.ts]: save `new Date(timestamp)` to a variable and reuse it instead of call `new Date` twice.

File: src/user/room.ts (hunk 1 above)
Comment (windmemory) [reply src/user/room.ts]: Will do.

File: src/wechaty.ts (hunk 2 above)
Comment (windmemory) [reply src/wechaty.ts]: This is aligning the `dong` event to all the other events, since this change modified the indentation of other events, is this necessary?

Comment (huan) [reply src/wechaty.ts]: Not necessary for this PR, because we have to focus on the code design.

We can fix the align after this PR had been merged by another PR, or we add the align as the lastest commit after this PR had been approved and ready to be merged.

File: src/wechaty.ts (hunk 3 above)
Comment (windmemory) [reply src/wechaty.ts]: I would prefer to have different variables, since users might modify the value of the variable(you never know).
If they listened on `wechat`'s `room-join` events, also listened to the specific `room`'s `room-join` event, modifying one time will have effect on the other one, which might cause some very unexpected bug in the future.

Alternative solution: we could make this value readonly or freeze this time variable.

What do you think about this?

File: src/wechaty.ts (hunk 2 above)
Comment (windmemory) [reply src/wechaty.ts]: okay, will revert this change.

**Comment by huan** [comment]
@windmemory Could you please mark my reviews as `resolved` after you resolve it? Or they will keep showing and it will not very clear for our progress. Thanks.

File: src/user/room.ts [hunk 4]
```diff
@@ -544,9 +544,9 @@ export class Room extends Accessory implements Sayable {
   }
 
   public emit (event: 'invite', inviter: Contact,         invitation: RoomInvitation)           : boolean
-  public emit (event: 'leave',  leaverList:   Contact[],  remover?: Contact)                    : boolean
-  public emit (event: 'join',   inviteeList:  Contact[],  inviter:  Contact)                    : boolean
-  public emit (event: 'topic',  topic:        string,     oldTopic: string,   changer: Contact) : boolean
+  public emit (event: 'leave',  leaverList:   Contact[],  remover:  Contact, time?: Date)                    : boolean
```
Comment (huan) [inline src/user/room.ts]: The `time` should be optional for `emit` because we must make sure if there's some consumer want to get the time, they will.

Remove `?` to make sure the `time` will be passed.

And I also feel that the `time` should rename to `date` because it's a `Date`.

File: src/wechaty.ts [hunk 5]
```diff
@@ -295,9 +295,9 @@ export class Wechaty extends Accessory implements Sayable {
   public emit (event: 'message',    message: Message)                                                   : boolean
   public emit (event: 'ready')                                                                          : boolean
   public emit (event: 'room-invite',  roomInvitation: RoomInvitation)                                   : boolean
-  public emit (event: 'room-join',    room: Room, inviteeList : Contact[], inviter  : Contact)          : boolean
-  public emit (event: 'room-leave',   room: Room, leaverList  : Contact[], remover? : Contact)          : boolean
-  public emit (event: 'room-topic',   room: Room, newTopic: string, oldTopic: string, changer: Contact) : boolean
+  public emit (event: 'room-join',    room: Room, inviteeList : Contact[], inviter : Contact, time?: Date)           : boolean
+  public emit (event: 'room-leave',   room: Room, leaverList  : Contact[], remover : Contact, time?: Date)           : boolean
```
Comment (huan) [inline src/wechaty.ts]: 1. Should be optional
1. Consider rename it to `date`

File: src/wechaty.ts [hunk 6]
```diff
@@ -683,27 +683,24 @@ export class Wechaty extends Accessory implements Sayable {
             const inviter = this.Contact.load(inviterId)
             await inviter.ready()
 
-            this.emit('room-join', room, inviteeList, inviter)
-            room.emit('join', inviteeList, inviter)
+            this.emit('room-join', room, inviteeList, inviter, new Date(timestamp))
```
Comment (huan) [inline src/wechaty.ts]: Create a `date` var to store the `new Date()` value. 

Then you can use it for twice.

File: src/wechaty.ts [hunk 7]
```diff
@@ -683,27 +683,24 @@ export class Wechaty extends Accessory implements Sayable {
...
-            if (removerId) {
-              remover = this.Contact.load(removerId)
-              await remover.ready()
-            }
+            const remover = this.Contact.load(removerId)
+            await remover.ready()
 
-            this.emit('room-leave', room, leaverList, remover)
-            room.emit('leave', leaverList, remover)
+            this.emit('room-leave', room, leaverList, remover, new Date(timestamp))
```
Comment (huan) [inline src/wechaty.ts]: Create a temp variable from here as well.

File: src/wechaty.ts (hunk 3 above)
Comment (huan) [reply src/wechaty.ts]: I don't think this will happen because the user can only modify their local copy of the variable value.

File: src/user/room.ts (hunk 4 above)
Comment (windmemory) [reply src/user/room.ts]: actually it is a `datetime`, shall we rename it as `datetime`?

**Comment by windmemory** [comment]
Revised code according to review and one comment to discuss.

File: src/user/room.ts [hunk 8]
```diff
@@ -583,9 +583,9 @@ export class Room extends Accessory implements Sayable {
   }
 
   public on (event: 'invite', listener: (this: Room, inviter: Contact, invitation: RoomInvitation) => void)               : this
-  public on (event: 'leave',  listener: (this: Room, leaverList:  Contact[], remover?: Contact) => void)                  : this
-  public on (event: 'join',   listener: (this: Room, inviteeList: Contact[], inviter:  Contact) => void)                  : this
-  public on (event: 'topic',  listener: (this: Room, topic:       string,    oldTopic: string, changer: Contact) => void) : this
+  public on (event: 'leave',  listener: (this: Room, leaverList:  Contact[], remover:  Contact, date: Date) => void)                  : this
```
Comment (huan) [inline src/user/room.ts]: We have to keep the `remover?` as optional or we will introduce breaking changes to our library.

And so that you will have to make the `date?` optional too.

File: src/wechaty.ts [hunk 9]
```diff
@@ -322,9 +322,9 @@ export class Wechaty extends Accessory implements Sayable {
...
   public on (event: 'ready',        listener: string | ((this: Wechaty) => void))                                                                   : this
   public on (event: 'room-invite',  listener: string | ((this: Wechaty, roomInvitation: RoomInvitation) => void))                                   : this
-  public on (event: 'room-join',    listener: string | ((this: Wechaty, room: Room, inviteeList: Contact[],  inviter: Contact) => void))            : this
-  public on (event: 'room-leave',   listener: string | ((this: Wechaty, room: Room, leaverList: Contact[], remover?: Contact) => void))             : this
-  public on (event: 'room-topic',   listener: string | ((this: Wechaty, room: Room, newTopic: string, oldTopic: string, changer: Contact) => void)) : this
+  public on (event: 'room-join',    listener: string | ((this: Wechaty, room: Room, inviteeList: Contact[], inviter: Contact, date?: Date) => void))             : this
+  public on (event: 'room-leave',   listener: string | ((this: Wechaty, room: Room, leaverList: Contact[],  remover: Contact, date?: Date) => void))             : this
```
Comment (huan) [inline src/wechaty.ts]: optional.

I would like to make the `remover` not optional, in the next major version upgrade.

**Comment by windmemory** [comment]
That's right, changed according to review suggestions.

File: package.json [hunk 10]
```diff
@@ -98,7 +98,7 @@
     "semver": "^6.0.0",
     "state-switch": "^0.6.2",
     "watchdog": "^0.8.1",
-    "wechaty-puppet": "^0.15.25",
+    "wechaty-puppet": "^0.15.26",
```
Comment (huan) [inline package.json]: Do you want to upgrade this version as well?

Comment (windmemory) [reply package.json]: Yes, I should have.

Bumped the version to `0.15.27`

**Comment by huan** [comment]
Merged.
